### PR TITLE
[Inductor][CPP] Enable fusion of BMM into onednn linear_pointwise

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139055

**Summary**
Here is the case in LLama2, Linear will be decomposed into BMM when input is 3D, non-contiguous and without bias. Here in this PR, we fuse it back into `onednn linear_pointwise` to speedup performance.

**TestPlan**
```
python -u -m pytest -s -v test/inductor/test_mkldnn_pattern_matcher.py -k test_linear_input_non_contiguous_3D_wo_bias
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/fx_passes/mkldnn_fusion.py [hunk 1]
```diff
@@ -945,6 +952,29 @@ def _is_packable_linear(match):
         weight_meta_value = linear_node.args[weight_idx].meta.get("val")
         if input_meta_value is None or weight_meta_value is None:
             return False
+        if linear_node.target == aten.bmm.default:
+            input_size = input_meta_value.size()
+            weight_size = weight_meta_value.size()
+            # bmm must be 3D tensor
+            assert len(input_size) == 3 and len(weight_size) == 3
+            if input_size[0] != weight_size[0]:
```
Comment (jgong5) [inline torch/_inductor/fx_passes/mkldnn_fusion.py:960]: If it is bmm, the first dim should always match? Should be assertion here?

File: torch/_inductor/fx_passes/mkldnn_fusion.py [hunk 2]
```diff
@@ -945,6 +952,29 @@ def _is_packable_linear(match):
...
+                return False
+            if input_size[-1] != weight_size[1]:
```
Comment (jgong5) [inline torch/_inductor/fx_passes/mkldnn_fusion.py:962]: ditto: should be assertion here?

File: torch/_inductor/fx_passes/mkldnn_fusion.py [hunk 3]
```diff
@@ -945,6 +952,29 @@ def _is_packable_linear(match):
...
+            bmm_wgt = getattr(
+                match.graph.owning_module, linear_node.args[weight_idx].name
+            )
+            with torch.utils._python_dispatch._disable_current_modes():
+                dim_0_size = bmm_wgt.size(0)
+                slice_0_val = bmm_wgt[0:1, :, :]
+                if not all(
+                    torch.allclose(bmm_wgt[i : (i + 1), :, :], slice_0_val)
+                    for i in range(dim_0_size)
+                ):
```
Comment (jgong5) [inline torch/_inductor/fx_passes/mkldnn_fusion.py:975]: Can we just check the tensor stride here?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/mkldnn_fusion.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/mkldnn_fusion.py:960]: Thanks for the comment and changed.

File: torch/_inductor/fx_passes/mkldnn_fusion.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/_inductor/fx_passes/mkldnn_fusion.py:962]: changed.

**Comment by leslie-fang-intel** [comment]
Sync with Jiong offline for different solutions, reopen in https://github.com/pytorch/pytorch/pull/139172
